pkg/havener/kubexec.go

```diff
@@ -115,83 +116,69 @@ func PodExec(client kubernetes.Interface, restconfig *rest.Config, pod *corev1.P
 	// TODO These fields should be made customizable using a configuration file
-	namespace := "kube-system"
-	containerName := "runon"
-	containerImage := "debian:jessie"
+	var err error
+
+	nodes, err := ListNodes(client)
+	if err != nil {
+		return err
+	}
+	if !sliceContainsString(nodes, node) {
+		return fmt.Errorf("invalid node: node '%s' does not exist\n\nAvailable nodes:\n%s",
+			node,
+			strings.Join(nodes, "\n"),
+		)
+	}
 
-	jobName := strings.ToLower(text.RandomStringWithPrefix("node-runner-", 24))
+	namespace := "kube-system"
+	containerName := text.RandomStringWithPrefix("node-exec-", 15) // Create unique pod/container name
 	trueThat := true
-	jobDef := &batchv1.Job{
+
+	// Pod confoguration
+	pod := &corev1.Pod{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      jobName,
+			Name:      containerName,
 			Namespace: namespace,
 		},
-		Spec: batchv1.JobSpec{
-			Template: corev1.PodTemplateSpec{
-				Spec: corev1.PodSpec{
-					NodeSelector:  map[string]string{"kubernetes.io/hostname": node},
-					HostPID:       true,
-					RestartPolicy: corev1.RestartPolicyNever,
-					Containers: []corev1.Container{
-						{
-							Name:    containerName,
-							Image:   containerImage,
-							Command: []string{"sleep", "8h"},
-							SecurityContext: &corev1.SecurityContext{
-								Privileged: &trueThat,
-							},
-						},
+		Spec: corev1.PodSpec{
+			NodeSelector:  map[string]string{"kubernetes.io/hostname": node}, // Deploy pod on specific node using label selector
+			HostPID:       true,
+			RestartPolicy: corev1.RestartPolicyNever,
+			Containers: []corev1.Container{
+				{
+					Name:  containerName,
+					Image: containerImage,
+					SecurityContext: &corev1.SecurityContext{
+						Privileged: &trueThat,
 					},
+					Stdin: true,
 				},
 			},
 		},
 	}
 
-	job, err := client.BatchV1().Jobs(namespace).Create(jobDef)
+	// Create pod in given namespace based on configuration
+	pod, err = client.CoreV1().Pods(namespace).Create(pod)
 	if err != nil {
 		return err
 	}
-
-	// Make sure that both the job and the pod it spawned are removed
-	// from the clusters once this function reaches its end.
+	// Stop pod and container after command execution
 	defer func() {
 		jobDeletionGracePeriod := int64(10)
 		propagationPolicy := metav1.DeletePropagationForeground
-		client.BatchV1().Jobs(namespace).Delete(job.Name, &metav1.DeleteOptions{
+		client.CoreV1().Pods(namespace).Delete(containerName, &metav1.DeleteOptions{
 			GracePeriodSeconds: &jobDeletionGracePeriod,
 			PropagationPolicy:  &propagationPolicy,
 		})
 	}()
 
-	pod, err := getPodUsingListOptions(client, namespace, metav1.ListOptions{LabelSelector: fmt.Sprintf("controller-uid=%s,job-name=%s", job.ObjectMeta.UID, jobName)})
-	if err != nil {
-		return err
-	}
-
-	// The reference to the pod that was spawned for the job.
-	watcher, err := client.CoreV1().Pods(namespace).Watch(metav1.SingleObject(pod.ObjectMeta))
+	err = waitForPodReadiness(client, namespace, pod, timeoutSeconds)
```

Comment:
In my opinion, inline if-clauses are less readable because the actual condition can be overseen very easily and it looks more like a variable assignment but on the other hand they have the advantage that they make the code more compact